File: Components/Strata.Hello.Client/main/java/strata/hello/client/requestreply/HelloServiceClient.java
//////////////////////////////////////////////////////////////////////////////
// HelloServiceClient.java
//////////////////////////////////////////////////////////////////////////////

package strata.hello.client.requestreply;

import jakarta.ws.rs.client.ClientBuilder;
import org.apache.hc.core5.ssl.SSLContextBuilder;
import org.springframework.core.io.ClassPathResource;
import org.springframework.core.io.Resource;
import strata.client.core.service.AbstractRestClient;
import strata.hello.service.requestreply.IHelloService;
import strata.hello.service.requestreply.SayHelloReply;
import strata.hello.service.requestreply.SayHelloRequest;

import javax.net.ssl.SSLContext;
import java.util.concurrent.CompletionStage;


public
class HelloServiceClient
    extends AbstractRestClient
    implements IHelloService
{

    public
    HelloServiceClient(String baseUrl)
    {
        super(
            createBuilder(),
            preprocessBaseUrl(baseUrl),
            "hello-service/");

    }

    @Override
    public SayHelloReply
    sayHelloSync(SayHelloRequest request)
    {
        return
            doPostAsync("say-hello-sync",SayHelloReply.class,request)
                .toCompletableFuture()
                .join();
    }

    @Override
    public CompletionStage<SayHelloReply>
    sayHelloAsync(SayHelloRequest request)
    {
        return
            doPostAsync("say-hello-async",SayHelloReply.class,request);
    }

    private static String
    preprocessBaseUrl(String baseUrl)
    {
        return
            baseUrl.endsWith("/")
                ? baseUrl
                : baseUrl + '/';
    }

    private static ClientBuilder
    createBuilder() throws RuntimeException
    {
        return
            ClientBuilder
                .newBuilder()
                .sslContext(createSslContext());
    }

    private static SSLContext
    createSslContext() throws RuntimeException
    {
        try
        {
            Resource keystore = new ClassPathResource("hello-service.p12");
            return
                new SSLContextBuilder()
                    .loadTrustMaterial(
                        keystore.getURL(),
                        "hello-service".toCharArray())
                    .build();
        }
        catch (Exception e)
        {
            throw new RuntimeException(e);
        }

    }
}

//////////////////////////////////////////////////////////////////////////////

File: Components/Strata.Hello.Client/main/typescript/RequestReply/HelloServiceClient.ts
import {IHelloService} from "strata.hello.service/RequestReply";
import {SayHelloRequest} from "strata.hello.service/RequestReply";
import {SayHelloReply} from "strata.hello.service/RequestReply";
import {ICompletionStage} from "strata.foundation.core/Concurrent";
import {AbstractRestClient} from "strata.client.core/Service";
import {Guid} from "guid-typescript";
import {Instant} from "@js-joda/core";

export
class HelloServiceClient
    extends AbstractRestClient
    implements IHelloService
{
    constructor(baseUrl: string)
    {
        super(baseUrl);
        this.setHeader("Content-Type","application/json");
    }

    sayHelloAsync(request: SayHelloRequest): ICompletionStage<SayHelloReply>
    {
        console.log("HelloServiceClient.sayHelloAsync");
        return this.doPost<SayHelloReply,SayHelloRequest>(
            "say-hello-async",
            this.initializeRequest(request));
    }

    private initializeRequest(request: SayHelloRequest): SayHelloRequest
    {
        console.log("HelloServiceClient.initializeRequest");
        if (request == null)
            request.requestId = Guid.create().toString();

        if (request.timestamp == null)
            request.timestamp = Instant.now().toString();

        return request;
    }
}